components/eam/src/chemistry/modal_aero/modal_aero_drydep.F90

```diff
@@ -0,0 +1,736 @@
+    type(cam_out_t),        intent(inout) :: cam_out   ! export state
+    type(physics_ptend),    intent(out)   :: ptend     ! indivdual parameterization tendencies
+    type(physics_buffer_desc),    pointer :: pbuf(:)
+
+  ! local vars
+    real(r8), pointer :: landfrac(:) ! land fraction
+    real(r8), pointer :: icefrac(:)  ! ice fraction
+    real(r8), pointer :: ocnfrac(:)  ! ocean fraction
+    real(r8), pointer :: fvin(:)     !
+    real(r8), pointer :: ram1in(:)   ! for dry dep velocities from land model for progseasalts
+
+    real(r8), pointer :: tair(:,:)   ! air temperture [k]
+    real(r8), pointer :: pmid(:,:)   ! air pressure at layer midpoint [Pa]
+    real(r8), pointer :: pint(:,:)   ! air pressure at layer interface [Pa]
+    real(r8), pointer :: pdel(:,:)   ! layer thickness [Pa]
+
+    real(r8) :: fv(pcols)            ! for dry dep velocities, from land modified over ocean & ice
+    real(r8) :: ram1(pcols)          ! for dry dep velocities, from land modified over ocean & ice
+
+    integer :: lchnk   ! chunk identifier
+    integer :: ncol    ! number of active atmospheric columns
+    integer :: lspec   ! index for aerosol number / chem-mass / water-mass
+    integer :: imode   ! aerosol mode index
+    integer :: icnst   ! tracer index
+    integer :: imnt    ! moment of the aerosol size distribution. 0 = number; 3 = volume
+    integer :: jvlc    ! index for last dimension of vlc_xxx arrays
+
+    real(r8) :: rho(pcols,pver)      ! air density in kg/m3
+    real(r8) :: sflx(pcols)          ! deposition flux
+    real(r8) :: dqdt_tmp(pcols,pver) ! temporary array to hold tendency for 1 species
+
+    real(r8) :: rad_drop(pcols,pver)
+    real(r8) :: dens_drop(pcols,pver)
+    real(r8) :: sg_drop(pcols,pver)
+    real(r8) :: rad_aer(pcols,pver)
+    real(r8) :: dens_aer(pcols,pver)
+    real(r8) :: sg_aer(pcols,pver)
+
+    real(r8) :: vlc_dry(pcols,pver,4)     ! dep velocity [m/s]
+    real(r8) :: vlc_grv(pcols,pver,4)     ! dep velocity [m/s]
+    real(r8)::  vlc_trb(pcols,4)          ! dep velocity [m/s]
+
+    real(r8) :: aerdepdryis(pcols,pcnst)  ! aerosol dry deposition (interstitial)
+    real(r8) :: aerdepdrycw(pcols,pcnst)  ! aerosol dry deposition (cloud water)
+
+    real(r8), pointer :: qq(:,:)            ! mixing ratio of a single tracer [kg/kg] or [1/kg]
+    real(r8), pointer :: dgncur_awet(:,:,:)
+    real(r8), pointer :: wetdens(:,:,:)
+
+    !---------------------------------------------------------------------------
+    ! Retrieve input variables; initialize output (i.e., ptend).
+    !---------------------------------------------------------------------------
+    landfrac => cam_in%landfrac(:)
+    icefrac  => cam_in%icefrac(:)
+    ocnfrac  => cam_in%ocnfrac(:)
+    fvin     => cam_in%fv(:)
+    ram1in   => cam_in%ram1(:)
+
+    lchnk = state%lchnk
+    ncol  = state%ncol
+
+    tair => state%t(:,:)
+    pmid => state%pmid(:,:)
+    pint => state%pint(:,:)
+    pdel => state%pdel(:,:)
```

Comment:
We might need to change the current way calling state variables. Since C++ port won't use state structure to store data, our current way is to make them as subroutine input.

Reply:
Hi @mingxuanwupnnl and @tangshuaiqi, I've thought a bit more about the comment above and wanted to ask for your opinions on the following.

The subroutine we are discussing here is the "main" subroutine of aerosol dry deposition in EAM, and this same subroutine also doubles as the interface to EAM's "physics driver" (`tphysac`+`tphysbc`). I can easily remove the lines marked above outside this subroutine and pass the individual quantities (arrays and scalars) instead, but that brings up a few follow-up questions:

- Would you like to have `call physics_ptend_init(...)` moved outside this subroutine, too?
- How about those `call pbuf_get_field(...)` statements?
- Later in the subroutine, inside the mode+species loops, we also have `qq => qqcw_get_field(pbuf,icnst,lchnk)` and `qq => state%q(:,:,icnst)`; what is the preferred way of handling those?

Perhaps what we could choose to do is to not change those lines in the Fortran code, but add comments (notes) to the C++ team that these should be replaced by their way of interfacing with the AD/aerosol-interface, like what you have done for some other aerosol processes?

Thanks!

Reply:
Thanks for bringing up these points.  If I may add some comments / questions (and realizing there are still some aspects of the current and future E3SM structure I am learning:

*)  At Balwinder's suggestion, I've been trying to move dereferencing elements of the state structure, including interstitial aerosols, towards the top of the subroutine hierarchy, which so far for me is 'dropmixnuc' in ndrop.F90.  I could continue the process one more level to microp_aero_run, so state would not need to be passed to dropmixnuc -- but then of course would need to coordinate with whoever is refactoring microp_aero_run.  

*)  But my understanding is that, regardless of the above, pbuf needs to be passed down through the subroutines, and ptend received upwards through the subroutines, that compute the physics tendencies.  So I am not sure what the advantage would be to moving these upward in the hierarchy (though if these are not moved upward while all references to the state structure are moved upward, then 'lchnk' and 'psetcols' would need to be extracted and passed down as arguments along with all the other needed elements of state).

*)  For interstitial aerosols in 'dropmixnuc', I have been generally using direct references to the state q vector, like the second option above.  But since, according to my understanding, cloud-borne aerosols only live in pbuf, I use the qqcw_get_field syntax to retrieve those from pbuf and lchnk, so the first option.  It sounds like we could use the 'qqcw_get_field' interface for both interstitial and cloud-borne aerosols, in which case the respective codes would appear more similar.  But the direct 'state' references seem to be more standard for t, pint, etc., and these also seem more readable / understandable to my eyes.

Reply:
@huiwanpnnl This is something I am not 100% sure of, and thanks @brigaudet for the valuable thoughts. 

My understanding is that we try to avoid `state` and `ptend` data structure in the low-level subroutines, by extracting the individual variables from these structures and passing to the child subroutines. `ptend` can also be passed as individual tendency variables and updated in the top-most subroutine. I am not quite sure about `pbuf`, looks that we can also do similar thing, or is it any exceptions?

However, so far we keep all the FORTRAN structure variables in the top-level subroutines such as drivers and interfaces. I don't know how the C++ team treat these variables. If your code is a main driver I think it is Okay to keep it as it. I like your idea of comment it and let C++ team decide if they can port it in their way or if they need us to modify it.

Reply:
@huiwanpnnl Since aero_model_drydep is already the driver subroutine which is called in tphysac, I think we don't need to change these lines but just leave a note for C++ port.

Reply:
@huiwanpnnl We may want to have two notes one for state variables and one for cam_in. C++ version will have different coupler array structure? Have not met cam_in input before for the refactoring work.